src/SimpleMemSystem.cpp

```diff
@@ -73,148 +75,205 @@ uint32_t printMessage(std::string key, std::string message, uint32_t maxCount)
     return count;
 }
 
-void AccessError(uint64_t pc, uint64_t addr, bool isWrite)
-{
-    std::stringstream msg;
-    msg << "Simulated code at address " << std::hex << pc << " tried to "
-        << (isWrite ? "write" : "read") << " unavailable memory at " << addr;
-    etiss::log(etiss::ERROR, msg.str());
+void SimpleMemSystem::init_memory() {
+    load_segments();
+    load_elf();
+    std::sort(msegs_.begin(), msegs_.end(), [](std::unique_ptr<MemSegment> & a, std::unique_ptr<MemSegment> & b) {return a->start_addr_ < b->start_addr_;});
 }
 
-etiss::int8 SimpleMemSystem::load_elf(const char *elf_file)
+void SimpleMemSystem::load_segments() {
+    std::stringstream ss;
+
+    for (int i = 0; i < MAX_MEMSEGS; ++i) {
+        ss << "simple_mem_system.memseg_origin_" << std::setw(2) << std::setfill('0') << i;
+        uint64_t origin = etiss::cfg().get<uint64_t>(ss.str(), -1);
+        std::stringstream().swap(ss);
+
+        ss << "simple_mem_system.memseg_length_" << std::setw(2) << std::setfill('0') << i;
+        uint64_t length = etiss::cfg().get<uint64_t>(ss.str(), -1);
+        std::stringstream().swap(ss);
+
+        ss << "simple_mem_system.memseg_image_" << std::setw(2) << std::setfill('0') << i;
+        std::string image = etiss::cfg().get<std::string>(ss.str(), "");
+        std::stringstream().swap(ss);
+
+        ss << "simple_mem_system.memseg_mode_" << std::setw(2) << std::setfill('0') << i;
+        std::string mode = etiss::cfg().get<std::string>(ss.str(), "");
+        std::stringstream().swap(ss);
+
+        if (origin != (etiss::uint64) -1 && length != (etiss::uint64) -1) {
+            configured_address_spaces_[origin] = length;
+
+            int access = MemSegment::UNSET;
+            std::string modestr = "";
+            if (mode.find('R') != mode.npos) {
+                access |= MemSegment::READ;
+                modestr += "R";
+            }
+            if (mode.find('W') != mode.npos) {
+                access |= MemSegment::WRITE;
+                modestr += "W";
+            }
+            if (mode.find('X') != mode.npos) {
+                access |= MemSegment::EXEC;
+                modestr += "X";
+            }
+
+            std::stringstream sname;
+            sname << i + 1 << " - " << modestr
+                << "[0x" << std::hex << std::setfill('0') << std::setw(sizeof(etiss::uint64) * 2) << origin + length - 1 << " - "
+                << "0x" << std::hex << std::setfill('0') << std::setw(sizeof(etiss::uint64) * 2) << origin << "]";
+
+            etiss::uint8 *buf = nullptr;
+            size_t fsize = 0;
+
+            if (image != "")
+            {
+                std::ifstream ifs;
```

Comment:
some style comments:

- prefer constructor over `.open`
- `if (!ifs)` is sufficient here and easier to read
- `.close` is not needed here

Reply:
Resolved with 3d186072ee1e0d1a556927d5f0169002fffc03cb